loadgen/kitchensink/helpers.go

```diff
@@ -0,0 +1,139 @@
+	Client     client.Client
+	WorkflowID string
+	RunID      string
+}
+
+func (e *ClientActionsExecutor) ExecuteClientSequence(ctx context.Context, clientSeq *ClientSequence) error {
+	println("!!!!!!!! Number of action sets ", len(clientSeq.ActionSets))
+	for _, actionSet := range clientSeq.ActionSets {
+		if err := e.executeClientActionSet(ctx, actionSet); err != nil {
+			return err
+		}
+		println("!!!!!!!!! Done")
+	}
+
+	return nil
+}
+
+func (e *ClientActionsExecutor) executeClientActionSet(ctx context.Context, actionSet *ClientActionSet) error {
+	println("!!!!!!!! Running action set is concurrent: ", actionSet.Concurrent)
+	errs, errGroupCtx := errgroup.WithContext(ctx)
+	for _, action := range actionSet.Actions {
+		if actionSet.Concurrent {
+			action := action
+			errs.Go(func() error {
+				err := e.executeClientAction(errGroupCtx, action)
+				if err != nil {
+					return fmt.Errorf("failed to execute client action %v: %w", action, err)
+				}
+				return nil
+			})
+		} else {
+			if err := e.executeClientAction(ctx, action); err != nil {
+				return fmt.Errorf("failed to execute client action %v: %w", action, err)
+			}
+		}
+	}
+	if actionSet.Concurrent {
+		if err := errs.Wait(); err != nil {
+			return err
+		}
+		if actionSet.WaitAtEnd != nil {
+			println("WAITING AT END", actionSet.WaitAtEnd.Seconds, " n ", actionSet.WaitAtEnd.Nanos)
+			select {
+			case <-time.After(actionSet.WaitAtEnd.AsDuration()):
+				println("Done waiting")
+			case <-ctx.Done():
+				return fmt.Errorf("context done while waiting for end %w", ctx.Err())
+			}
+		}
+	}
+	return nil
+}
+
+// Run a specific client action -
+func (e *ClientActionsExecutor) executeClientAction(ctx context.Context, action *ClientAction) error {
+	if action.Variant == nil {
+		return fmt.Errorf("client action variant must be set")
+	}
+
+	println("Running client action: ", proto.MarshalTextString(action))
+
+	var err error
+	if sig := action.GetDoSignal(); sig != nil {
+		if actionSet := sig.GetDoActions(); actionSet != nil {
+			err = e.Client.SignalWorkflow(ctx, e.WorkflowID, e.RunID, "do_actions_signal", actionSet)
+		} else if handler := sig.GetCustom(); handler != nil {
+			err = e.Client.SignalWorkflow(ctx, e.WorkflowID, e.RunID, handler.Name, handler.Args)
+		} else {
+			return fmt.Errorf("do_signal must recognizable variant")
+		}
+	} else if update := action.GetDoUpdate(); update != nil {
+		if actionsUpdate := update.GetDoActions(); actionsUpdate != nil {
+			_, err = e.Client.UpdateWorkflow(ctx, e.WorkflowID, e.RunID, "do_actions_update", actionsUpdate)
+		} else if handler := update.GetCustom(); handler != nil {
+			_, err = e.Client.UpdateWorkflow(ctx, e.WorkflowID, e.RunID, handler.Name, handler.Args)
+		} else {
+			return fmt.Errorf("do_update must recognizable variant")
+		}
+		if update.FailureExpected {
+			err = nil
+		}
+	} else if query := action.GetDoQuery(); query != nil {
+		if query.GetReportState() != nil {
+			// TODO: Use args
```

Comment:
Do you want an issue for this?

Reply:
Nah, I'll just come in one of the follow up PRs where I add more features